File: python/package/robot_simulator/robot_2DOF.py
import numpy as np
import matplotlib.pyplot as plt
import math

## elbow_up = 1
## elbow_down = 2

class robot_2DOF:
    def __init__(self, name, length, bias, elbow_case):
        self.name = name
        self.link1_length = length[0]
        self.link2_length = length[1]
        self.biasx = bias[0]
        self.biasy = bias[1]
        self.elbow_case = elbow_case

    def fkine(self, q):
        l1 = self.link1_length
        l2 = self.link2_length
        q1 = q[0]
        q2 = q[1]
        ##
        p1x = l1 * np.cos(q1) + self.biasx
        p1y = l1 * np.sin(q1) + self.biasy
        p2x = l1 * np.cos(q1) + l2 * np.cos(q1 + q2) + self.biasx
        p2y = l1 * np.sin(q1) + l2 * np.sin(q1 + q2) + self.biasy

        posi = [p1x, p1y, p2x, p2y]
        return np.array(posi)

    def isField(self, p2):
        l1 = self.link1_length
        l2 = self.link2_length
        x = p2[0] - self.biasx
        y = p2[1] - self.biasy
        ##
        judge = (np.square(x) + np.square(y) - np.square(l1) - np.square(l2)) / (2 * l1 * l2)
        if np.abs(judge) > 1:
            isField = False
        else:
            isField = True
        return isField

    def ikine(self, p2):
        l1 = self.link1_length
        l2 = self.link2_length
        x = p2[0] - self.biasx
        y = p2[1] - self.biasy
        ##
        isField = self.isField(p2)
        if isField == False:
            q = []
            print('The position is not in the workspace, please try again')
            return
        ##
        c2 = (np.square(x) + np.square(y) - np.square(l1) - np.square(l2)) / (2 * l1 * l2)

        if self.elbow_case == 1:
            s2 = -np.sqrt(1 - np.square(c2))
        elif self.elbow_case == 2:
            s2 = np.sqrt(1 - np.square(c2))
        else:
            print('Elbow case wrong, please try again')
            return

        q2 = math.atan2(s2, c2)
        k1 = l1 + l2 * c2
        k2 = l2 * s2
        q1 = math.atan2(y, x) - math.atan2(k2, k1)
        q = [q1, q2]
        return np.array(q)

    def plot(self, q):
        l1 = self.link1_length
        l2 = self.link2_length
        ##
        posi = self.fkine(q)
        p1x = posi[0]
        p1y = posi[1]
        p2x = posi[2]
        p2y = posi[3]

        plt.cla()
        plt.plot([self.biasx, p1x, p2x], [self.biasy, p1y, p2y], color='black', linestyle='-', lw=2, marker='*')
        axis_limit = l1 + l2 + max(self.biasx, self.biasy)
        plt.axis('square')
        plt.xlim((-0.2, axis_limit))
        plt.ylim((-0.2, axis_limit))


if __name__ == '__main__':
    length = [1, 1]
    bias = [0.2, 0.3]
    robot = robot_2DOF('robot', length, bias, 2)

    plt.ion()
    plt.show()

    N = 200
    theta = np.linspace(0, 2 * np.pi, N)

    traj = []
    for i in range(N):
        traj.append([0.7 + 0.3 * math.cos(theta[i]), 0.8 + 0.3 * math.sin(theta[i])])
    traj = np.array(traj)

    for i in range(N):
        p2 = [traj[i, 0], traj[i, 1]]
        q = robot.ikine(p2)
        robot.plot(q)
        plt.plot(traj[1:i, 0], traj[1:i, 1], linestyle='--', color='red', lw=2)
        plt.xlim((-1.5, 1.5))
        plt.ylim((-1.5, 1.5))
        plt.pause(0.01)
        ##
        rad = q.sum() - np.pi
        print(np.rad2deg(rad), '  ', robot.fkine(q))

File: matlab/package/robot_simulator_2D/programme_class/robot_2DOF.m
classdef robot_2DOF
    %UNTITLED Summary of this class goes here
    %   Detailed explanation goes here
    
    properties
        link1_length;
        link2_length;
        biasx;
        biasy;
        elbow_case; %up = 1 and down = 2
    end
    
    methods
        function obj = robot_2DOF(link1,link2,biasx,biasy,elbow_case)
            %UNTITLED Construct an instance of this class
            %   Detailed explanation goes here
            obj.link1_length = link1;
            obj.link2_length = link2;
            obj.biasx = biasx;
            obj.biasy = biasy;
            obj.elbow_case = elbow_case;
        end
    
        function [p1,p2] = fkine(obj,q)
           %% initial parameter
            l1 = obj.link1_length;
            l2 = obj.link2_length;
            q1 = q(1);
            q2 = q(2);
            bias.x = obj.biasx;
            bias.y = obj.biasy;
           %%
            p1.x = l1 * cos(q1) + bias.x;
            p1.y = l1 * sin(q1) + bias.y;
            p2.x = l1 * cos(q1) + l2 * cos(q1 + q2) + bias.x;
            p2.y = l1 * sin(q1) + l2 * sin(q1 + q2) + bias.y;
        end
        
        function isField = isField(obj,p2)
           %% initial parameter
            l1 = obj.link1_length;
            l2 = obj.link2_length;
            bias.x = obj.biasx;
            bias.y = obj.biasy;
           %%
            x = p2.x - bias.x;
            y = p2.y - bias.y;
            judge = (x^2 + y^2 -l1^2 -l2^2)/(2*l1*l2);
            if abs(judge) > 1
                isField = false;
            else
                isField = true;
            end
        end
        
        function q = ikine(obj,p2)
           %% initial parameter
            l1 = obj.link1_length;
            l2 = obj.link2_length;
            bias.x = obj.biasx;
            bias.y = obj.biasy;
           %%
            x = p2.x - bias.x;
            y = p2.y - bias.y;
            isField = obj.isField(p2);
            if ~isField
                q = [];
                fprintf('The position is not in the workspace, pleast try it again\n')
                return;
            end 
      
           %%
            c2 = (x^2 + y^2 -l1^2 -l2^2)/(2*l1*l2);
            switch obj.elbow_case
                case 1
                    s2 = -sqrt(1-c2^2);
                case 2
                    s2 = sqrt(1-c2^2);
            end
            q2 = atan2(s2,c2);
            k1 = l1 + l2 * c2;
            k2 = l2 * s2;
            q1 = atan2(y,x) - atan2(k2,k1);
            q = [q1;q2];       
        end
        
        function handle = plot(obj,q)
           %% initial parameter
            l1 = obj.link1_length;
            l2 = obj.link2_length;
            bias.x = obj.biasx;
            bias.y = obj.biasy;
           %%
           [p1,p2] = obj.fkine(q);
           %% plot figure
           handle = plot([bias.x p1.x p2.x],[bias.y p1.y p2.y],'k-*','linewidth',2);
           %% axis limitation
           axis_limit = l1 + l2 + max(bias.x,bias.y);
           axis([-0.2 axis_limit -0.2 axis_limit])
           daspect([1 1 1])
        end
        
        function circle(obj,radius,bias)
            n = 100;
            theta = linspace(0,2*pi,n);
            for i=1:n
                traj(i,1) = radius * cos(theta(i)) + bias.x;
                traj(i,2) = radius * sin(theta(i)) + bias.y;
            end
            plot(traj(:,1),traj(:,2),'--','linewidth',2)
        end
        
        function handle = rectangle(obj,p1x,p1y,p2x,p2y)
            %%
            % ---------- p2
            % |          |
            % |          |
            %p1 ---------
            %%
            p1.x = p1x;p1.y = p1y;p2.x = p2x;p2.y = p2y;
            %%
            handle = plot([p1.x p2.x p2.x p1.x p1.x],[p1.y p1.y p2.y p2.y p1.y],'k-','linewidth',2);
        end
        
        function handle = workspace(obj)
            maxlength = obj.link1_length + obj.link2_length + sqrt(obj.biasx^2 + obj.biasy^2);
            N = 200;
            theta = linspace(0,2*pi,N);
            for i=1:N
                x(i) = maxlength * cos(theta(i));
                y(i) = maxlength * sin(theta(i));
            end
            handle = plot(x,y,'--b','linewidth',2);
        end
        
    end
end

